pyro/infer/renyi_elbo.py

```diff
@@ -0,0 +1,179 @@
+        For :math:`\alpha = 1`, it is better to use
+        :class:`~pyro.infer.trace_elbo.Trace_ELBO` class because it helps reducing
+        variances of gradient estimations.
+
+    .. warning:: Mini-batch training is not supported yet.
+
+    :param float alpha: The order of :math:`\alpha`-divergence. Here
+        :math:`\alpha \neq 1`. Default is 0.
+    :param num_particles: The number of particles/samples used to form the ELBO
+        (gradient) estimators. Default is 2.
+    :param int max_iarange_nesting: Bound on max number of nested
+        :func:`pyro.iarange` contexts. Default is 2.
+    :param bool strict_enumeration_warning: Whether to warn about possible
+        misuse of enumeration, i.e. that
+        :class:`~pyro.infer.traceenum_elbo.TraceEnum_ELBO` is used iff there
+        are enumerated sample sites.
+
+    References:
+
+    [1] `Renyi Divergence Variational Inference`,
+        Yingzhen Li, Richard E. Turner
+
+    [2] `Importance Weighted Autoencoders`,
+        Yuri Burda, Roger Grosse, Ruslan Salakhutdinov
+    """
+
+    def __init__(self, alpha=0., num_particles=2, max_iarange_nesting=2,
+                 strict_enumeration_warning=True):
+        if alpha == 1:
+            raise ValueError("The order alpha should not be equal to 1. Please use Trace_ELBO class"
+                             "for the case alpha = 1.")
+        self.alpha = alpha
+        super(RenyiELBO, self).__init__(num_particles, max_iarange_nesting, vectorize_particles=True,
+                                        strict_enumeration_warning=True)
+
+    def _get_trace(self, model, guide, *args, **kwargs):
+        """
+        Returns a single trace from the guide, and the model that is run
+        against it.
+        """
+        guide_trace = poutine.trace(guide).get_trace(*args, **kwargs)
+        model_trace = poutine.trace(poutine.replay(model, trace=guide_trace)).get_trace(*args, **kwargs)
+        if is_validation_enabled():
+            check_model_guide_match(model_trace, guide_trace)
+            enumerated_sites = [name for name, site in guide_trace.nodes.items()
+                                if site["type"] == "sample" and site["infer"].get("enumerate")]
+            if enumerated_sites:
+                warnings.warn('\n'.join([
+                    'Trace_ELBO found sample sites configured for enumeration:'
+                    ', '.join(enumerated_sites),
+                    'If you want to enumerate sites, you need to use TraceEnum_ELBO instead.']))
+        guide_trace = prune_subsample_sites(guide_trace)
+        model_trace = prune_subsample_sites(model_trace)
+
+        model_trace.compute_log_prob()
+        guide_trace.compute_score_parts()
+        if is_validation_enabled():
+            for site in model_trace.nodes.values():
+                if site["type"] == "sample":
+                    check_site_shape(site, self.max_iarange_nesting)
```

Comment:
this is used verbatim in almost every *elbo implementation, we should eventually move this out to a helper